### PR TITLE
Update client path, docs, add additional methods

### PR DESCRIPTION
**Pull request by barbaraliau**

### CODE REVIEW

File: js/clients/README.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+```
+
+```js
+const { BcoinClient } = require('@summa-tx/bitcoin-spv-js-clients');
+```
+
+ES6 module:
+
+```js
+import { BconClient } from '@summa-tx/bitcoin-spv-js-clients';
```
Comment (prestwich) [inline js/clients/README.md]: bcon

File: js/clients/README.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+// [
+//   '8e2d404a039a7a3e1768b161aa23546aab0444b73905bdd3d68b3d6f1769e8c0...',
+//   4
+// ]
+```
+
+#### client.getHeaderChainByCount(height, count, enc)
+* `@param {Number} height` - starting block height
+* `@param {Number} count` - number of headers
+* `@param {String} enc` - json or hex
```
Comment (prestwich) [inline js/clients/README.md]: also supports `btcspv now`

File: js/clients/lib/BcoinClient.js [hunk 3]
```diff
@@ -143,6 +167,7 @@ class BcoinClient extends NodeClient {
    * @param {Number} height - starting block height
    * @param {Number} count - number of headers
    * @param {String} enc - json or hex
```
Comment (prestwich) [inline js/clients/lib/BcoinClient.js]: also `btcspv` now

File: js/clients/README.md (hunk 1 above)
Comment (barbaraliau) [reply js/clients/README.md]: akon

**Comment by barbaraliau** [comment]
Ok I made changes. And fixed the history.

File: js/clients/README.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+const client = new BcoinClient(options);
+const block = 12345;
+const header = client.getHeader(block);
+
+// {
+//   raw: hex,
+//   hash: json.hash,
+//   hash_le: reverse(json.hash),
+//   height: typeof block === 'number' ? block : json.height,
+//   prevhash: json.prevBlock,
```
Comment (prestwich) [inline js/clients/README.md]: this should include `prevhash_le` now as well

File: js/clients/lib/BcoinClient.js [hunk 5]
```diff
@@ -1,16 +1,23 @@
...
+* @copyright (c) [Summa]{@link https://summa.one} 2019
+* @module clients
+*
+*/
 
 'use strict';
 
-const {NodeClient} = require('./vendor/bclient');
+const { NodeClient } = require('./vendor/bclient');
 const assert = require('./vendor/bsert');
```
Comment (prestwich) [inline js/clients/lib/BcoinClient.js]: these two vendors are broken

File: js/clients/package.json [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  "version": "0.1.0",
+  "description": "Clients for fetching bitcoin SPV proofs in Javascript",
+  "main": "lib/index.js",
+  "engines": {
+    "node": ">10.0.0"
+  },
+  "engineStrict": true,
+  "scripts": {
+    "lint": "eslint ./test ./src",
+    "lint:fix": "eslint --fix ./test ./src",
```
Comment (prestwich) [inline js/clients/package.json]: these don't lint anything in this package

File: js/clients/package.json [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    "test": "mocha test -r esm",
+    "test:coverage": "jest --coverage",
```
Comment (prestwich) [inline js/clients/package.json]: no tests in this package

File: js/clients/lib/BcoinClient.js (hunk 5 above)
Comment (barbaraliau) [reply js/clients/lib/BcoinClient.js]: Fixed

File: js/clients/package.json (hunk 7 above)
Comment (barbaraliau) [reply js/clients/package.json]: Fixed

File: js/clients/package.json (hunk 6 above)
Comment (barbaraliau) [reply js/clients/package.json]: Fixed

**Comment by barbaraliau** [comment]
Realized that there are several variables that are not declared. Need to be ventured in.

**Comment by barbaraliau** [comment]
@prestwich variables come from using `bcoin` in `getHeaderChain(height, nwork)`. This is the method that was originally used when the `BcoinClient` was in `bitcoin-verifier-ses`.

Can this be replaced with `getHeaderChainByCount()`? This will change the use of predicates that use `nwork`.

**Comment by prestwich** [comment]
> Can this be replaced with getHeaderChainByCount()? This will change the use of predicates that use nwork.

unfortunately no. they have very different purposes

**Comment by barbaraliau** [comment]
Ok. This was the one method I didn't rewrite with no dependencies. I will
vendor or rewrite then.

**Comment by tynes** [comment]
If you lint the vendored deps, then its going to be a pain to diff them against new versions

File: js/clients/README.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+const header = client.getMerkleProof(txid, height);
+
+// merkle proof
+// [
+//   '8e2d404a039a7a3e1768b161aa23546aab0444b73905bdd3d68b3d6f1769e8c0...',
+//   4
+// ]
+```
+
+#### client.getHeaderChainByCount(height, count, enc)
```
Comment (tynes) [inline js/clients/README.md]: The concept of passing an encoding type is a little bit vestigial at this point, I originally had it because sometimes we wanted to pretty print and sometimes we wanted the raw array buffers. I think it may be a good idea to eliminate the `enc` variable to simplify and always return json (or uint8arrays if preferred) and then the ser package can be used to convert between the two.

File: js/clients/lib/BcoinClient.js [hunk 9]
```diff
@@ -19,12 +25,11 @@ const {utils, BTCUtils} = require('../js/dist');
  */
 class BcoinClient extends NodeClient {
   /**
-   * bcoinclient constructor.
+   * bcoinclient constructor
+   * NB: none since constructor doesn't do anything
+   * (see: eslint-no-useless-constructor)
    * @param {object} options
    */
-  constructor(options) {
```
Comment (tynes) [inline js/clients/lib/BcoinClient.js]: It will automatically call the `NodeClient` constructor and pass along the arguments? Didn't know that

File: js/clients/README.md (hunk 8 above)
Comment (barbaraliau) [reply js/clients/README.md]: Yeah that makes sense. I need to finish up this refactor and push all the changes I have.

File: js/clients/README.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+#### client.getHeader(block)
+* `@param {Hash|Number} block` - block height or hash
+* `@returns {Object} header` - header object
+
+Gets header by height or hash.
+```js
+const client = new BcoinClient(options);
+const block = 12345;
+const header = client.getHeader(block);
```
Comment (tynes) [inline js/clients/README.md]: Missing `await`

File: js/clients/README.md [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+#### client.getProof(txid)
+   * `@param {String} txid` - txid in big endian
+   * `@returns {Object}` - proof object
+
+Get a proof for a transaction.
+
+```js
+const client = new BcoinClient(options);
+const txid =
```
Comment (tynes) [inline js/clients/README.md]: Missing `txid` value

File: js/clients/README.md [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+//       "version": 536870912,
+//       "prevBlock": "0c4ea5e675941eca1909275f21903cef755069a03c57c10f4d4dadcdd7146daf",
+//       "merkleRoot": "9a249c682aba07943c8a1f9bd774a15d71372fcd7f3f9ee99e8c7aa022ae6aa0",
+//       "time": 1571661863,
+//       "bits": 545259519,
+//       "nonce": 8,
+//       "height": 100,
+//       "chainwork": "00000000000000000000000000000000000000000000000000000000000000ca"
+//     },
+//         {
```
Comment (tynes) [inline js/clients/README.md]: Spacing is off here

File: js/clients/lib/BcoinClient.js (hunk 9 above)
Comment (barbaraliau) [reply js/clients/lib/BcoinClient.js]: If there isn't one specified, it provides a default one. For derived classes:

```
constructor(...args) {
  super(...args);
}
```

I don't like unnecessary code, so it's kind of nice. If we decide we want to put them in at a later time then we can always disable the rule and add them back in.

File: js/clients/README.md (hunk 10 above)
Comment (barbaraliau) [reply js/clients/README.md]: Fixing.

Comment (barbaraliau) [reply js/clients/README.md]: And all the other ones.

**Comment by tynes** [comment]
Sorry about removing some of the methods that are needed for the ses code when I was porting, was only thinking about the CLI

**Comment by barbaraliau** [comment]
> If you lint the vendored deps, then its going to be a pain to diff them against different versions. Granted the vendored deps should be pretty stable, but you never know.

I added an `.eslintignore` so those files aren't linted. For the above reason, but also that would be a pain to lint lol. It was something like 323 linting errors.

**Comment by barbaraliau** [comment]
> Sorry about removing some of the methods that are needed for the ses code when I was porting, was only thinking about the CLI

@tynes No worries. I didn't realize it was needed either until I was rewriting the `README.md` for the ses code, and then realized that I needed to update all the `proof` objects to make sure that the client was returning the right type, since almost all of them need to be `Uint8Array`s

File: js/clients/README.md (hunk 8 above)
Comment (barbaraliau) [reply js/clients/README.md]: Going to return `Uint8Array`s for all of these since the primary use case is for the verifier code, and we can use the `ser` package to go back and forth as needed.

**Comment by tynes** [comment]
> @tynes No worries. I didn't realize it was needed either until I was rewriting the `README.md` for the ses code, and then realized that I needed to update all the `proof` objects to make sure that the client was returning the right type, since almost all of them need to be `Uint8Array`s

So the plan is to delete the client in the ses code and then use the one here?

**Comment by tynes** [comment]
Also see this commit https://github.com/summa-tx/bitcoin-spv/commit/559471b6c64a98ff42acdd1901b4266a2c77073f, I removed methods like `getHeadersByWork` and `getHeadersByCurrency`.

`getHeadersByWork` was meant to return an amount of headers that satisfies `nwork` (BN representation of work), summing each header's work together until surpasses `nwork`

Since thinking in `nwork` is not intuitive, `getHeadersByCurrency` was supposed to take a dollar amount and then use an external API to fetch the price of BTC and then try to calculate a dollar based amount of headers, so that you can say "this headers chain is 1 million dollars worth of security"

**Comment by barbaraliau** [comment]
> > @tynes No worries. I didn't realize it was needed either until I was rewriting the `README.md` for the ses code, and then realized that I needed to update all the `proof` objects to make sure that the client was returning the right type, since almost all of them need to be `Uint8Array`s
> 
> So the plan is to delete the client in the ses code and then use the one here?

TL;DR, yes. I was refactoring the client in the ses code, but then this code was pushed up and since we were, trying to remove as many dependencies as possible from the ses code, it made sense to finish this up and use it, since it was almost identical, barring one or two methods.

Side note: I wonder if we could pass in the preferred encoding when instantiating the client, as a nice way of saying I want all my methods to return this type, without having to specify it in each method, and use the `ser` package internally to do that.

File: js/clients/README.md (hunk 8 above)
Comment (barbaraliau) [reply js/clients/README.md]: Jk. I think I'll save this for a different time. For now use the `ser` package on the other side.

**Comment by tynes** [comment]
> Side note: I wonder if we could pass in the preferred encoding when instantiating the client, as a nice way of saying I want all my methods to return this type, without having to specify it in each method, and use the ser package internally to do that.

I like this idea, different applications would probably want the results in a particular format. The CLI wants all json and the ses code wants all uint8arrays

**Comment by prestwich** [comment]
Where are we on this?
